crates/recursor/src/recursor.rs

```diff
@@ -307,366 +301,106 @@ impl Recursor {
-/// | example.com. | www.example.com.     | In-bailiwick (true)                                              |
-/// | example.com. | www.otherdomain.com. | Out-of-bailiwick (false)                                         |
-/// | example.com  | www.example.com.     | Out-of-bailiwick (false, note the parent is not fully qualified) |
-///
-/// # Implementation Notes
-///
-/// * This function is nominally a wrapper around Name::zone_of, with two additional checks:
-/// * If the caller doesn't provide a parent at all, we'll return false.
-/// * If the domains have mixed qualification -- that is, if one is fully-qualified and the other partially-qualified, we'll return
-///    false.
-///
-/// # References
-///
-/// * [RFC 8499](https://datatracker.ietf.org/doc/html/rfc8499) -- DNS Terminology (see page 25)
-/// * [The Hitchiker's Guide to DNS Cache Poisoning](https://www.cs.utexas.edu/%7Eshmat/shmat_securecomm10.pdf) -- for a more in-depth
-/// discussion of DNS cache poisoning attacks, see section 4, specifically, for a discussion of the Bailiwick rule.
-fn is_subzone(parent: Name, child: Name) -> bool {
-    if parent.is_empty() {
-        return false;
-    }
-
-    if (parent.is_fqdn() && !child.is_fqdn()) || (!parent.is_fqdn() && child.is_fqdn()) {
-        return false;
+#[cfg(feature = "dnssec")]
+mod for_dnssec {
+    use std::time::Instant;
+
+    use futures_util::{
+        future,
+        stream::{self, BoxStream},
+        StreamExt as _, TryFutureExt as _,
+    };
+
+    use crate::proto::{
+        error::ProtoError, op::Message, op::OpCode, xfer::DnsHandle, xfer::DnsResponse,
+    };
+    use crate::recursor_dns_handle::RecursorDnsHandle;
+
+    impl DnsHandle for RecursorDnsHandle {
+        type Response = BoxStream<'static, Result<DnsResponse, ProtoError>>;
+
+        fn send<R: Into<hickory_proto::xfer::DnsRequest> + Unpin + Send + 'static>(
+            &self,
+            request: R,
+        ) -> Self::Response {
+            let request = request.into();
+
+            let query = if let OpCode::Query = request.op_code() {
+                if let Some(query) = request.queries().first().cloned() {
+                    query
+                } else {
+                    return Box::pin(stream::once(future::err(ProtoError::from(
+                        "no query in request",
+                    ))));
+                }
+            } else {
+                return Box::pin(stream::once(future::err(ProtoError::from(
+                    "request is not a query",
+                ))));
+            };
+
+            let this = self.clone();
+            stream::once(async move {
+                // request the DNSSEC records; we'll strip them if not needed on the caller side
+                let do_bit = true;
+                this.resolve(query, Instant::now(), do_bit)
+                    .map_ok(|lookup| {
+                        // XXX `DnssecDnsHandle` will only look at the answer section of the message so
```

Comment:
this implementation ended up being a bit "forced" as it assumes that this interface will only be used by the `DnssecDnsHandle` newtype, which is the case.

I feel like changing the `DnsHandle` interface would make *this* code reuse of `DnssecDnsHandle` cleaner but `DnsHandle` is used in other places so that would be a much larger change.

Similarly, I considered creating a new "lookup only" interface alternative to `DnsHandle` that would be used as a trait bound in `DnssecDnsHandle` but then I wasn't sure which of the other existing `DnsHandle` implementations should move to new interface and which others should keep using `DnsHandle`.